parsons/rockthevote/rtv.py

```diff
@@ -0,0 +1,211 @@
+                Parsons table with the report data.
+        """
+        credentials = {
+            'partner_id': self.partner_id,
+            'partner_API_key': self.partner_api_key,
+        }
+        status_url = f'{self.client.uri}/registrant_reports/{report_id}'
+        download_url = None
+
+        # Let's figure out at what time should we just give up because we waited
+        # too long
+        end_time = datetime.datetime.now() + datetime.timedelta(seconds=report_timeout_seconds)
+
+        # If we have a download URL, we can move on and just download the
+        # report. Otherwise, as long as we haven't run out of time, we will
+        # check the status.
+        while not download_url and datetime.datetime.now() < end_time:
+            logger.debug(f'Registrations report not ready yet, sleeping %s seconds',
+                         poll_interval_seconds)
+
+            # Check the status again via the status endpoint
+            status_response = self.client.request(status_url, 'get', params=credentials)
+
+            # Check to make sure the call got a valid response
+            if status_response.status_code == requests.codes.ok:
+                status_json = status_response.json()
+
+                # Grab the download_url from the response.
+                download_url = status_json.get('download_url')
+
+                if not download_url and not block:
+                    return None
+            else:
+                raise RTVFailure("Couldn't get report status")
+
+            if not download_url:
+                # We just got the status, so we should wait a minute before
+                # we check it again.
+                time.sleep(poll_interval_seconds)
+
+        # If we never got a valid download_url, then we timed out waiting for
+        # the report to generate. We will log an error and exit.
+        if not download_url:
+            raise RTVFailure('Timed out waiting for report')
+
+        # Download the report data
+        download_response = self.client.request(download_url, 'get', params=credentials)
+
+        # Check to make sure the call got a valid response
+        if download_response.status_code == requests.codes.ok:
+            report_data = download_response.text
+            with open('/tmp/rtv.csv', 'w') as fh:
+                fh.write(report_data)
+
+            # Load the report data into a Parsons Table
+            table = Table.from_csv_string(report_data)
+
+            # Transform the data from the report's CSV format to something more
+            # Pythonic (snake case)
+            normalized_column_names = [
+                re.sub(r'\s', '_', name).lower()
+                for name in table.columns
+            ]
+            normalized_column_names = [
+                re.sub(r'[^A-Za-z\d_]', '', name)
+                for name in normalized_column_names
+            ]
+            table.table = petl.setheader(table.table, normalized_column_names)
```

Comment:
Would love this as a `Table` class method. I feel like we could use this in many many places.

Reply:
Done.